backends/ebpf/psa/README.md

```diff
@@ -521,6 +521,114 @@ $ bpftool help
+}
+```
+
+Note that cached table entries are a copy of an original entry. This causes that extern which preserve internal state inside
+table entry (like `DirectMeter` or `DirectCounter`) can't be used with cache due to two existing different states. Compiler
+will not enable table caching for tables containing such direct extern.
+
+When updating table with enabled cache, the cache must be invalidated in order to update to be effective. The simplest way
+to do this is to remove all entries from LRU map, which is done automatically by the NIKSS library.
+
+Similar approach to this is done in `Action Selector` extern when group reference is present. Lookup with cache into
+`Action Selctor` looks like below where cache is used to skip checksum calculation and member selection if possible:
+```c
+struct ingress_as_value * as_value = NULL;  // pointer to an action data
+u32 as_action_ref = value->ingress_as_ref;  // value->ingress_as_ref is entry from table (reference)
+u8 as_group_state = 0;                      // from which map read action data
+struct ingress_as_key_cache key_cache = {0};
+u8 do_update_cache = 0;
+if (value->ingress_as_is_group_ref != 0) {
+    key_cache.group_ref = value->ingress_as_ref; // group reference
+    key_cache.field0 = /* fill selectors value */;
+    as_value = BPF_MAP_LOOKUP_ELEM(ingress_as_cache, &key_cache);
+    if (as_value != NULL) {
+        as_group_state = 2; // cache hit, forbid later lookups into maps
+    } else {
+        do_update_cache = 1; // cache miss, update cache later, below normal lookup
+        void * as_group_map = BPF_MAP_LOOKUP_ELEM(ingress_as_groups, &as_action_ref);  // get group map
+        if (as_group_map != NULL) {
+            u32 * num_of_members = bpf_map_lookup_elem(as_group_map, &ebpf_zero);
+            if (num_of_members != NULL) {
+                if (*num_of_members != 0) {
+                    /* calculate checksum here */
+                    u64 as_checksum_val = /* calculated checksum */;
+                    as_action_ref = /* determine member reference based on checksum */;
+                } else {
+                    as_group_state = 1; // execute default action when group is empty
+                }
+            } else {
+                return TC_ACT_SHOT; // number of members not found
+            }
+        } else {
+            return TC_ACT_SHOT; // group not found
+        }
+    }
+}
+if (as_group_state == 0) {
+    as_value = BPF_MAP_LOOKUP_ELEM(ingress_as_actions, &as_action_ref); // member action data (valid member reference)
+} else if (as_group_state == 1) {
+    as_value = BPF_MAP_LOOKUP_ELEM(ingress_as_defaultActionGroup, &ebpf_zero);  // default group action data
+}
+if (as_value != NULL && do_update_cache != 0) {
+    BPF_MAP_UPDATE_ELEM(ingress_as_cache, &key_cache, as_value, BPF_ANY); // update cache
+}
+```
+
 # TODO / Limitations
 
 We list the known bugs/limitations below. Refer to the Roadmap section for features planned in the near future.
```

Comment:
Can we add limitations of table caching to the list?

Reply:
I will add the limitations